nixos/modules/services/blockchain/ethereum/geth.nix

```diff
@@ -0,0 +1,198 @@
+            Port number of Go Ethereum WebSocket API.
+          '';
+        };
+
+        apis = mkOption {
+          type = types.nullOr (types.listOf types.str);
+          default = null;
+          description = ''
+            APIs to enable over WebSocket
+          '';
+          example = ["net" "eth"];
+        };
+      };
+
+      metrics = {
+        enable = lib.mkEnableOption "Go Ethereum prometheus metrics";
+        address = mkOption {
+          type = types.str;
+          default = "127.0.0.1";
+          description = ''
+            Listen address of Go Ethereum metrics service.
+          '';
+        };
+
+        port = mkOption {
+          type = types.port;
+          default = 6060;
+          description = ''
+            Port number of Go Ethereum metrics service.
+          '';
+        };
+      };
+
+      network = mkOption {
+        type = types.enum [ "mainnet" "goerli" "inkeby" "yolov2" "ropsten" ];
+        default = "mainnet";
+        description = ''
+          The network to connect to. Mainnet is the default ethereum network.
+        '';
+      };
+
+      syncmode = mkOption {
+        type = types.enum [ "fast" "full" "light" ];
+        default = "fast";
+        description = ''
+           Blockchain sync mode.
+        '';
+      };
+
+      gcmode = mkOption {
+        type = types.enum [ "full" "archive" ];
+        default = "full";
+        description = ''
+           Blockchain garbage collection mode.
+        '';
+      };
+
+      extraArgs = mkOption {
+        type = types.str;
+        description = ''
+          Additional arguments passed to Go Ethereum.
+        '';
+        default = "";
+      };
+
+      package = mkOption {
+        default = pkgs.geth;
+        type = types.package;
+        description = ''
+          Package to use as Go Ethereum node.
+        '';
+      };
+    };
+  };
+in
+
+{
+
+  ###### interface
+
+  options = {
+    services.geth = mkOption {
+      type = types.attrsOf (types.submodule gethOpts);
+      default = {};
+      description = "Specification of one or more geth instances.";
+    };
+  };
+
+  ###### implementation
+
+  config = mkIf (eachGeth != {}) {
+
+    systemd.services = mapAttrs' (gethName: cfg: (
+      nameValuePair "geth-${gethName}" (mkIf cfg.enable {
+      description = "Go Ethereum node (${gethName})";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+
+      serviceConfig = {
+        DynamicUser = true;
+        Restart = "always";
+        StateDirectory = "goethereum/${gethName}/${cfg.network}";
+
+        # Hardening measures
+        PrivateTmp = "true";
+        ProtectSystem = "full";
+        NoNewPrivileges = "true";
+        PrivateDevices = "true";
+        MemoryDenyWriteExecute = "true";
+      };
+
+      script = ''
+        ${cfg.package}/bin/geth \
+          --nousb \
+          --ipcdisable \
+          ${optionalString (cfg.network == "goerli") "--goerli"} \
```

Comment:
I'd make this one bif if statement which errors out on unknown networks, that will prevent any future programming mistakes in this module.

Reply:
Or maybe just simply `--${cfg.network}` as that seems OK for all currently defined networks.

Reply:
Or... Scratch that, this module should also be able to handle `cfg.network` being `null` which might be a reasonable thing for private networks.

Reply:
@adisbladis My proposal would be to change it to either: 
1. `types.nullOr types.string`
2. `types.nullOr types.enum [ "goerli" "rinkeby" "yolov2" "ropsten" ]`

Do you have a preference?

1. is more flexible as it allows any protential network without extending the service. 2. Is safer but requires to update the service in lockstep with the package.

Reply:
Per @adisbladis comment it makes sense to keep both service and package as is. As the geth service is just one think provided by the go-ethereum package.

Reply:
I think I prefer option 2. If you want to use a network which is not defined in the module you can always use `extraOptions`.

Reply:
It's now `types.nullOr (types.enum [ "goerli" "rinkeby" "yolov2" "ropsten" ])`